src/test/unit/mcmc/chainset_test.cpp

```diff
@@ -0,0 +1,334 @@
+  std::vector<stan::io::stan_csv> eight_schools_5iters;
+  eight_schools_5iters.push_back(eight_schools_5iters_1);
+  eight_schools_5iters.push_back(eight_schools_5iters_2);
+  stan::mcmc::chainset chain_2(eight_schools_5iters);
+  EXPECT_EQ(2, chain_2.num_chains());
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto ess = chain_2.split_rank_normalized_ess(i + 7);
+    EXPECT_TRUE(std::isnan(ess.first));
+    EXPECT_TRUE(std::isnan(ess.second));
+  }
+}
+
+TEST_F(McmcChains, summary_stats) {
+  std::vector<stan::io::stan_csv> eight_schools;
+  eight_schools.push_back(eight_schools_1);
+  eight_schools.push_back(eight_schools_2);
+  stan::mcmc::chainset chain_2(eight_schools);
+  EXPECT_EQ(2, chain_2.num_chains());
+
+  // test against R implementation in pkg posterior (via cmdstanr)
+  Eigen::VectorXd s8_mean(10), s8_median(10), s8_sd(10), s8_mad(10), s8_q5(10),
+      s8_q95(10);
+  s8_mean << 7.95, 12.54, 7.82, 5.33, 7.09, 4.12, 5.72, 11.65, 8.80, 8.26;
+  s8_median << 8.00, 11.27, 7.39, 5.44, 6.64, 4.54, 5.93, 11.38, 8.28, 7.05;
+  s8_sd << 5.48, 9.57, 6.85, 8.39, 6.91, 6.57, 6.85, 7.76, 8.40, 5.53;
+  s8_mad << 5.49, 8.79, 6.39, 7.38, 5.98, 6.25, 6.59, 7.79, 7.59, 4.66;
+  s8_q5 << -0.46, -0.39, -3.04, -8.90, -3.31, -7.58, -5.84, 0.10, -4.15, 2.08;
+  s8_q95 << 17.01, 30.47, 19.25, 19.02, 18.72, 14.49, 16.04, 25.77, 22.71,
+      18.74;
+  Eigen::VectorXd probs(3);
+  probs << 0.05, 0.5, 0.95;
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto mean = chain_2.mean(i + 7);
+    EXPECT_NEAR(mean, s8_mean(i), 0.05);
+    auto median = chain_2.median(i + 7);
+    EXPECT_NEAR(median, s8_median(i), 0.05);
+    auto sd = chain_2.sd(i + 7);
+    EXPECT_NEAR(sd, s8_sd(i), 0.05);
+    auto mad = chain_2.max_abs_deviation(i + 7);
+    EXPECT_NEAR(mad, s8_mad(i), 0.05);
+    auto q_5 = chain_2.quantile(i + 7, 0.05);
+    EXPECT_NEAR(q_5, s8_q5(i), 0.5);
+    auto q_95 = chain_2.quantile(i + 7, 0.95);
+    EXPECT_NEAR(q_95, s8_q95(i), 0.5);
+    auto qs_5_50_95 = chain_2.quantiles(i + 7, probs);
+    EXPECT_EQ(3, qs_5_50_95.size());
+    EXPECT_NEAR(qs_5_50_95(0), s8_q5(i), 0.5);
+    EXPECT_NEAR(qs_5_50_95(1), s8_median(i), 0.05);
+    EXPECT_NEAR(qs_5_50_95(2), s8_q95(i), 0.5);
+  }
+}
+
+TEST_F(McmcChains, mcse) {
+  std::vector<stan::io::stan_csv> eight_schools;
+  eight_schools.push_back(eight_schools_1);
+  eight_schools.push_back(eight_schools_2);
+  stan::mcmc::chainset chain_2(eight_schools);
+  EXPECT_EQ(2, chain_2.num_chains());
+
+  // test against R implementation in pkg posterior
+  Eigen::VectorXd s8_mcse_mean(10), s8_mcse_sd(10);
+  s8_mcse_mean << 0.288379, 0.4741815, 0.2741001, 0.3294614, 0.2473758,
+      0.2665048, 0.2701363, 0.4740092, 0.3621771, 0.3832464;
+  s8_mcse_sd << 0.1841825, 0.2854258, 0.192332, 0.2919369, 0.2478025, 0.2207478,
+      0.2308452, 0.2522107, 0.2946896, 0.3184745;
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto mcse_mean = chain_2.mcse_mean(i + 7);
+    EXPECT_NEAR(mcse_mean, s8_mcse_mean(i), 0.5);
```

Comment:
Again -- absolute tolerance is larger than most values in this array we're testing against

I think all the EXPECT_NEARs in this file need to be tightened, some replaced with EXPECT_FLOAT_EQ